src/main/scala/mesosphere/marathon/state/Group.scala

```diff
@@ -272,23 +285,63 @@ object Group extends StrictLogging {
       // this is funny: id is "optional" only because version and scaleBy can't be used in conjunction with other
       // fields. it feels like we should make an exception for "id" and always require it for non-root groups.
       group.id.map(_.toPath) as "id" is optional(valid)
-
-      group.apps is optional(every(
-        AppValidation.validNestedApp(group.id.fold(base)(PathId(_).canonicalPath(base)))))
-      group.groups is optional(every(
-        validNestedGroupUpdateWithBase(group.id.fold(base)(PathId(_).canonicalPath(base)), originalRootGroup)))
     }
 
-  private case class definingEnforcingRoleOnlyIfItsTopLevel(base: AbsolutePathId) extends Validator[raml.GroupUpdate] {
-    override def apply(group: raml.GroupUpdate): Result = {
-      val groupId = group.id.fold(base) { id => PathId(id).canonicalPath(base) }
-      // Only top-level groups are allowed to set the enforce role parameter.
-      if (!groupId.isTopLevel && group.enforceRole.contains(true)) {
-        Failure(Set(RuleViolation(group.enforceRole, s"enforceRole can only be set for top-level groups, and ${groupId} is not top-level", Path(Generic("enforceRole")))))
+    def enforceRoleNotDefined(thisGroup: GroupUpdate, groupId: AbsolutePathId): Result = {
+      if (thisGroup.enforceRole.contains(true)) {
+        Failure(Set(RuleViolation(thisGroup.enforceRole, s"enforceRole can only be set for top-level groups, and $groupId is not top-level", Path(Generic("enforceRole")))))
       } else {
         Success
       }
     }
+
+    override def visit(thisGroup: raml.GroupUpdate): Result = groupUpdateValidator(thisGroup)
+
+    override def appVisitor(): AppVisitor[Result] = AppValidationVisitor()
+
+    override def childGroupVisitor(): GroupValidationVisitor = this
+  }
+
+  case class TopLevelGroupValidationVisitor(override val originalRootGroup: RootGroup) extends GroupValidationVisitor {
+
+    override def childGroupVisitor(): GroupValidationVisitor = LowerLevelGroupValidationVisitor(originalRootGroup)
+  }
+
+  case class RootGroupValidationVisitor(override val originalRootGroup: RootGroup) extends GroupValidationVisitor {
+
+    override def visit(thisGroup: GroupUpdate): Result = super.visit(thisGroup).and(enforceRoleNotDefined(thisGroup, PathId.root))
+
+    override def childGroupVisitor(): GroupValidationVisitor = TopLevelGroupValidationVisitor(originalRootGroup)
+  }
+
+  case class LowerLevelGroupValidationVisitor(override val originalRootGroup: RootGroup) extends GroupValidationVisitor {
+
+    override def visit(thisGroup: GroupUpdate): Result = super.visit(thisGroup).and(enforceRoleNotDefined(thisGroup, PathId.root)) // TODO: pass group path
```

Comment:
Bug! We need to pass the correct absolute group id. However, the visit method does not have the parameter yet.